test/parallel-benchmark/mzcompose.py

```diff
@@ -0,0 +1,1190 @@
+
+    def run(
+        self,
+        c: Composition,
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        print(f"Load phase for {self.duration}s")
+        threads = [
+            threading.Thread(
+                target=phase_action.run,
+                args=(self.duration, jobs, conns, measurements),
+            )
+            for phase_action in self.phase_actions
+        ]
+        for thread in threads:
+            thread.start()
+        for i in range(self.duration):
+            assert (
+                jobs.qsize() < 100
+            ), "Too many jobs in queue, can't keep up: {jobs.qsize()}"
+            # if jobs.qsize() != 0:
+            #    measurements["Job queue"].append(Measurement(jobs.qsize()))
+            # if conns.qsize() != 0:
+            #    measurements["Connection queue"].append(Measurement(conns.qsize()))
+            time.sleep(1)
+        for thread in threads:
+            thread.join()
+
+
+def run_job(jobs: queue.Queue) -> None:
+    while True:
+        job = jobs.get()
+        try:
+            if not job:
+                return
+            job()
+        finally:
+            jobs.task_done()
+
+
+class Scenario:
+    phases: list[Phase]
+    thread_pool_size: int
+    conn_pool_size: int
+    guarantees: dict[str, dict[str, float]]
+    jobs: queue.Queue
+    conns: queue.Queue
+    thread_pool: list[threading.Thread]
+
+    def __init__(self, c: Composition, conn_infos: dict[str, PgConnInfo]):
+        raise NotImplementedError
+
+    def init(
+        self,
+        phases: list[Phase],
+        thread_pool_size: int = 5000,
+        conn_pool_size: int = 0,
+        guarantees: dict[str, dict[str, float]] = {},
+    ):
+        self.phases = phases
+        self.thread_pool_size = thread_pool_size
+        self.conn_pool_size = conn_pool_size
+        self.guarantees = guarantees
+        self.jobs = queue.Queue()
+        self.conns = queue.Queue()
+
+    def setup(self, c: Composition, conn_infos: dict[str, PgConnInfo]) -> None:
+        conn_info = conn_infos["materialized"]
+        self.thread_pool = [
+            threading.Thread(target=run_job, args=(self.jobs,))
+            for i in range(self.thread_pool_size)
+        ]
+        for thread in self.thread_pool:
+            thread.start()
+        # Start threads and have them wait for work from a queue
+        for i in range(self.conn_pool_size):
+            self.conns.put(conn_info.connect())
+
+    def run(
+        self,
+        c: Composition,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        for phase in self.phases:
+            phase.run(c, self.jobs, self.conns, measurements)
+
+    def teardown(self) -> None:
+        while not self.conns.empty():
+            conn = self.conns.get()
+            conn.close()
+            self.conns.task_done()
```

Comment:
Is this correctly indented or does it need to be outside of the while block?

Reply:
It is correct!